### PR TITLE
Added EcoBytes struct

### PR DESCRIPTION
**Pull request by imbrem**
Add `EcoBytes`, which is the same as `EcoString` but holds a raw byte vector rather than a UTF-8 string; this is implemented by making `DynamicVec` public.

### CODE REVIEW

**Comment by imbrem** [comment]
@reknih any updates on whether I can expect this to be merged? Thanks!

File: src/bytes.rs [hunk 1]
```diff
@@ -0,0 +1,749 @@
...
+        };
+        EcoBytes::from_inline(inline)
+    }
+
+    /// Create a new, inline vector.
+    ///
+    /// Returns `Err(())` if the vector's length exceeds the capacity of
+    /// the inline storage.
+    #[inline]
+    pub const fn try_inline(bytes: &[u8]) -> Result<Self, ()> {
```
Comment (laurmaedje) [inline src/bytes.rs:116]: I thinking returning `Option` would be preferrable over having `()` as an error.

File: src/bytes.rs [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+
+    /// Create a new, empty vector.
+    #[inline]
+    pub const fn new() -> Self {
+        Self::from_inline(InlineVec::new())
+    }
+
+    /// Create a new, inline vector.
+    ///
+    /// Panics if the vector's length exceeds the capacity of the inline
```
Comment (laurmaedje) [inline src/bytes.rs:101]:
```suggestion
    /// Panics if the slice's length exceeds the capacity of the inline
```

File: src/bytes.rs [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+    pub const fn try_inline(bytes: &[u8]) -> Result<Self, ()> {
+        match InlineVec::from_slice(bytes) {
+            Ok(inline) => Ok(EcoBytes::from_inline(inline)),
+            _ => Err(()),
+        }
+    }
+
+    #[inline]
+    pub(crate) const fn from_inline(inline: InlineVec) -> Self {
+        Self(Repr { inline })
```
Comment (laurmaedje) [inline src/bytes.rs:125]: this function doesn't really pull it's weight in my opinion

File: src/bytes.rs [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+        unsafe {
+            // Safety: self.len().saturating_sub(1) <= self.len()
+            self.set_len(len);
+        }
+        last
+    }
+
+    #[inline]
+    pub fn insert(&mut self, index: usize, value: u8) -> Result<(), ()> {
+        //TODO: optimize?
```
Comment (laurmaedje) [inline src/bytes.rs:627]: I think a custom implementation with `ptr::copy` (like in vec.rs) does make sense here.

File: src/bytes.rs [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+        let len = slice.len();
+        if index >= len {
+            out_of_bounds(index, len - 1)
+        }
+        slice[index..].rotate_right(1);
+        Ok(())
+    }
+
+    #[inline]
+    pub fn remove(&mut self, index: usize) -> u8 {
```
Comment (laurmaedje) [inline src/bytes.rs:639]: see above (like in vec.rs)

**Review by laurmaedje — COMMENTED** [review]
Sorry for the long wait. PRs outside typst/typst sometimes fly under my radar.

**Comment by laurmaedje** [comment]
Could you also adjust the README and crate-level docs and briefly show `EcoBytes`?

**Comment by laurmaedje** [comment]
Closing this due to inactivity. Feel free to reopen it if you want to continue working on this.

**Pull request closed without merging**